src/piton/labelers/core.py

```diff
@@ -351,17 +268,142 @@ def __len__(self):
+
+        Returns:
+            List[Label]: A list of :class:`Label` containing every label for the given patient
+        """
+        pass
+
+    def get_patient_start_end_times(
+        self, patient: Patient
+    ) -> Tuple[datetime.datetime, datetime.datetime]:
+        """Return the (start, end) of the patient timeline.
+
+        Returns:
+            Tuple[datetime.datetime, datetime.datetime]: (start, end)
+        """
+        return (patient.events[0].start, patient.events[-1].start)
+
+    @abstractmethod
+    def get_labeler_type(self) -> LabelType:
+        """Return what type of labels this labeler returns. See the Label class."""
+        pass
+
+    def apply(
+        self,
+        path_to_patient_database: Optional[str] = None,
+        patients: Optional[Sequence[Patient]] = None,
+        num_threads: int = 1,
+        num_patients: Optional[int] = None,
+    ) -> LabeledPatients:
+        """Apply the `label()` function one-by-one to each Patient in a sequence of Patients.
+
+        Args:
+            path_to_patient_database (str, optional): Path to `PatientDatabase` on disk.
+                Must be specified if `patients = None`
+            patients (Sequence[Patient], optional): An Sequence (i.e. list) of `Patient` objects.
+                Must be specified if `path_to_patient_database = None`
+                Typically this will be a `PatientDatabase` object.
+            num_threads (int, optional): Number of CPU threads to parallelize across. Defaults to 1.
+            num_patients (Optional[int], optional): Number of patients to process - useful for debugging.
+                If specified, will take the first `num_patients` in the provided `PatientDatabase` / `patients` list.
+                If None, use all patients.
+
+        Returns:
+            LabeledPatients: Maps patients to labels
+        """
+        if (patients is None and path_to_patient_database is None) or (
+            patients is not None and path_to_patient_database is not None
+        ):
+            raise ValueError(
+                "Must specify exactly one of `patient_database` or `path_to_patient_database`"
+            )
+
+        if path_to_patient_database:
+            # Load patientdatabase if specified
+            assert patients is None
+            patient_database = PatientDatabase(path_to_patient_database)
+            num_patients = (
+                len(patient_database) if not num_patients else num_patients
+            )
+            pids = list(range(num_patients))
+        else:
+            # Use `patients` if specified
+            assert patients is not None
+            num_patients = len(patients) if not num_patients else num_patients
+            pids = [p.patient_id for p in patients[:num_patients]]
+
+        # Split patient IDs across parallelized processes
+        pids_parts = np.array_split(pids, num_threads)
+
+        # NOTE: Super hacky workaround to pickling limitations
```

Comment:
I am not sure I understand this workaround. Why do we need this hack?

Reply:
If an object has a `self.ontology` property (which has type `extensions.Ontology`), we won't be able to pickle that object (i.e. `self`) since `extensions.Ontology` is not pickleable. Thus, we won't be able to pass the labeler (i.e. `self`) to multiprocessing b/c it's not pickable